src/elfcore.cc

```diff
@@ -1255,26 +1449,81 @@ static int CreateElfCore(void *handle, ssize_t (*writer)(void *, const void *, s
-        for (i = 0; i < num_mappings; i++) {
-          if (mappings[i].write_size > 0 &&
-              writer(handle, (void *)mappings[i].start_address, mappings[i].write_size) != mappings[i].write_size) {
+        if (ext_process) {
+          /* Write all memory segments                                         */
+          char read_buf[pagesize];
+          sprintf(proc_path, "/proc/%d/mem", main_pid);
+          NO_INTR(mem_fd = sys_open(proc_path, O_RDONLY, 0));
+          if (mem_fd == -1) {
             goto done;
           }
-        }
-        if (vdso.address) {
-          /* Finally write the contents of Phdrs that "belong" to vdso.      */
-          Phdr *vdso_phdr = (Phdr *)(vdso.address + vdso.ehdr->e_phoff);
-          for (i = 0; i < vdso.ehdr->e_phnum; i++) {
-            Phdr *p = vdso_phdr + i;
-            if (p->p_type == PT_LOAD) {
-              /* This segment has already been dumped, because it is one of
-               * the mappings[].
-               */
-            } else if (writer(handle, (void *)p->p_vaddr, p->p_filesz) != p->p_filesz) {
+          for (i = 0; i < num_mappings; i++) {
+            if (mappings[i].write_size > 0) {
+                size_t pending = mappings[i].write_size;
+                size_t inter;
+                size_t nread;
+                if (sys_lseek(mem_fd, mappings[i].start_address, SEEK_SET) < 0) {
+                    sys_close(mem_fd);
+#ifdef DEBUG
+                    fprintf(stderr, "Cannot Seek to address %lx err=%d\n",
+                                    mappings[i].start_address, errno);
+#endif
+                    goto done;
+                }
+                while (pending > 0) {
+                    inter = MIN(pending, pagesize);
+                    NO_INTR(nread = sys_read(mem_fd, read_buf, inter));
+                    if (writer(handle, (void *)read_buf, nread) != nread) {
+                        sys_close(mem_fd);
+                        goto done;
+                    }
+                    pending -= nread;
+                }
+            }
+          }
+          sys_close(mem_fd);
+
+          if (vdso.address) {
+            /* Finally write the contents of Phdrs that "belong" to vdso.      */
+            for (i = 0; i < vdso.ehdr.e_phnum; i++) {
+              Phdr *p = vdso_phdr + i;
+              if (p->p_type == PT_LOAD) {
+                /* This segment has already been dumped, because it is one of
+                 * the mappings[].
+                 */
+              } else {
+                char scratch[p->p_filesz];
+
+                if (PeekVDSO((void *)p->p_vaddr, (void *)scratch, p->p_filesz, main_pid) < 0) {
+                  goto done;
+                }
+                if (writer(handle, (void *)scratch, p->p_filesz) != p->p_filesz) {
+                  goto done;
+                }
+              }
+            }
+          }
+        } else { // self core write
```

Comment:
why is this all different? Can we commonize better?

Reply:
No, we cannot. In case of external process we are reading /proc/[pid]/mem to read the memory, and using local address in case of self.